### PR TITLE
PartitionKeyDelete: Adds API to delete all items in a partition key

### PR DESCRIPTION
**Pull request by aayush3011**
Added deleteAllItemsByPartitionKeyAsync API

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosContainer.java [hunk 1]
```diff
@@ -632,6 +632,10 @@ private CosmosBatchResponse blockBatchResponse(Mono<CosmosBatchResponse> batchRe
         return  this.blockDeleteItemResponse(asyncContainer.deleteItem(itemId, partitionKey, options));
     }
 
+    public CosmosItemResponse<Object> deleteAllItemsByPartitionKeyAsync(PartitionKey partitionKey, CosmosItemRequestOptions options) {
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosContainer.java:635]: `deleteAllItemsByPartitionKeyAsync` or `deleteAllItemsByPartitionKey`?

please check other APIs in this class, follow the same naming pattern

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java [hunk 2]
```diff
@@ -1473,6 +1475,15 @@ public String getId() {
         return withContext(context -> deleteItemInternal(itemId, null, requestOptions, context));
     }
 
+    public Mono<CosmosItemResponse<Object>> deleteAllItemsByPartitionKeyAsync(PartitionKey partitionKey, CosmosItemRequestOptions options) {
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:1478]: `deleteAllItemsByPartitionKeyAsync` or `deleteAllItemsByPartitionKey`?

please check other APIs in this class, follow the same naming pattern

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/BaseAuthorizationTokenProvider.java [hunk 3]
```diff
@@ -49,6 +49,7 @@ private static String getResourceSegment(ResourceType resourceType) {
         case Document:
             return Paths.DOCUMENTS_PATH_SEGMENT;
         case DocumentCollection:
+        case PartitionKey:
```
Comment (simplynaveen20) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/BaseAuthorizationTokenProvider.java:52]: is this intentional to return colls path for PartitionKey ?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/PathsHelper.java [hunk 4]
```diff
@@ -63,14 +63,16 @@ public static String generatePathForNameBased(Resource resourceType, String reso
...
         if (isFeed && Strings.isNullOrEmpty(resourceFullName) && resourceType != ResourceType.Database) {
             String errorMessage = String.format(RMResources.UnexpectedResourceType, resourceType);
             throw new IllegalArgumentException(errorMessage);
         }
 
         String resourcePath = null;
-        if (!isFeed) {
+        if (resourceType == ResourceType.PartitionKey && operationType == OperationType.Delete) {
+            resourcePath = resourceFullName + "/" + Paths.OPERATIONS_PATH_SEGMENT + "/" + Paths.PARTITION_KEY_DELETE_PATH_SEGMENT;
+        } else if (!isFeed) {
```
Comment (simplynaveen20) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/PathsHelper.java:75]: I would keep the isFeed condition first , as it has separate logic then rest of all, easier to read

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java [hunk 5]
```diff
@@ -1941,6 +1953,42 @@ void captureSessionToken(RxDocumentServiceRequest request, RxDocumentServiceResp
...
+
+            Mono<Utils.ValueHolder<DocumentCollection>> collectionObs = collectionCache.resolveCollectionAsync(BridgeInternal.getMetaDataDiagnosticContext(request.requestContext.cosmosDiagnostics), request);
+
+            Mono<RxDocumentServiceRequest> requestObs = addPartitionKeyInformation(request, null, null, options, collectionObs);
+
+            return requestObs.flatMap(req -> this
+                .deleteAllItemsByPartitionKey(req, retryPolicyInstance, getOperationContextAndListenerTuple(options))
+                .map(serviceResponse -> toResourceResponse(serviceResponse, Document.class)));
+        } catch (Exception e) {
+            logger.debug("Failure in deleting a document due to [{}]", e.getMessage());
```
Comment (simplynaveen20) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java:1987]: nit: Failure in deleting a document -> Failure in deleting documents

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxGatewayStoreModel.java [hunk 6]
```diff
@@ -109,6 +109,10 @@ public RxGatewayStoreModel(
         return this.performRequest(request, HttpMethod.DELETE);
     }
 
+    private Mono<RxDocumentServiceResponse> deleteByPartitionKey(RxDocumentServiceRequest request) {
+        return this.performRequest(request, HttpMethod.POST);
```
Comment (simplynaveen20) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxGatewayStoreModel.java:113]: Is this need to be post ?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/BaseAuthorizationTokenProvider.java (hunk 3 above)
Comment (simplynaveen20) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/BaseAuthorizationTokenProvider.java:52]: You might need to added support for resource token requests as it is introducing new resource type , please check method https://github.com/Azure/azure-sdk-for-java/blob/c28d9a2e4b40feece2a26799d0fecb31d6721b84/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/BaseAuthorizationTokenProvider.java#L177
to verify you can add a simple test in this test class https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/ResourceTokenTestForV4.java

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/PathsHelper.java [hunk 7]
```diff
@@ -109,7 +111,7 @@ private static String generatePathForNameBased(ResourceType resourceType, String
         return resourcePath;
     }
 
-    public static String generatePath(ResourceType resourceType, String ownerOrResourceId, boolean isFeed) {
+    public static String generatePath(ResourceType resourceType, String ownerOrResourceId, boolean isFeed, OperationType operationType) {
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/PathsHelper.java:114]: if you add an overload `PathHelper.generatePath` with operationType = null default. then you don't have to modify many other files.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/ChangeFeedFetcher.java [hunk 8]
```diff
@@ -62,7 +62,7 @@ public ChangeFeedFetcher(
         } else {
             DocumentClientRetryPolicy retryPolicyInstance = client.getResetSessionTokenRetryPolicy().getRequestPolicy();
             String collectionLink = PathsHelper.generatePath(
-                ResourceType.DocumentCollection, changeFeedState.getContainerRid(), false);
+                ResourceType.DocumentCollection, changeFeedState.getContainerRid(), false, null);
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/ChangeFeedFetcher.java:65]: if you add an overload `PathHelper.generatePath` with operationType = null default. then you don't have to modify many other files.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/BaseAuthorizationTokenProvider.java (hunk 3 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/BaseAuthorizationTokenProvider.java:52]: Yes, the .net PR returns the colls path for PartitionKey

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxGatewayStoreModel.java (hunk 6 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxGatewayStoreModel.java:113]: Yes, according to the design doc, it has to go as POST.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosContainer.java (hunk 1 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosContainer.java]: Replaced with deleteAllItemsByPartitionKey

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java (hunk 2 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java]: Replaced with deleteAllItemsByPartitionKey

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java (hunk 5 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java]: Done

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/PathsHelper.java (hunk 4 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/PathsHelper.java]: Done

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/ChangeFeedFetcher.java (hunk 8 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/ChangeFeedFetcher.java]: Done

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/PathsHelper.java (hunk 7 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/PathsHelper.java]: Done

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosContainer.java [hunk 9]
```diff
@@ -632,6 +632,10 @@ private CosmosBatchResponse blockBatchResponse(Mono<CosmosBatchResponse> batchRe
         return  this.blockDeleteItemResponse(asyncContainer.deleteItem(itemId, partitionKey, options));
     }
 
+    public CosmosItemResponse<Object> deleteAllItemsByPartitionKey(PartitionKey partitionKey, CosmosItemRequestOptions options) {
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosContainer.java:635]: javadoc is missing.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java [hunk 10]
```diff
@@ -1473,6 +1475,16 @@ public String getId() {
         return withContext(context -> deleteItemInternal(itemId, null, requestOptions, context));
     }
 
+    @Beta(value = Beta.SinceVersion.V4_19_0, warningText = Beta.PREVIEW_SUBJECT_TO_CHANGE_WARNING)
+    public Mono<CosmosItemResponse<Object>> deleteAllItemsByPartitionKey(PartitionKey partitionKey, CosmosItemRequestOptions options) {
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:1479]: javadoc is missing.

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosItemTest.java [hunk 11]
```diff
@@ -155,6 +155,52 @@ public void deleteItem() throws Exception {
         assertThat(deleteResponse.getStatusCode()).isEqualTo(204);
     }
 
+    @Test(groups = {"simple"}, timeOut = TIMEOUT)
+    public void deleteAllItemsByPartitionKeyAsync() throws Exception {
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosItemTest.java:159]: I don't see anything async in this method. any reason this is suffixed. with async?

if it is not testing async api, please remove the async suffix

**Review by moderakh — APPROVED** [review]
Thanks @aayush3011

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosContainer.java (hunk 9 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosContainer.java:645]: Added

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java (hunk 10 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:1489]: Added

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosItemTest.java (hunk 11 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosItemTest.java]: Removed the async suffix.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosContainer.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosContainer.java:645]: The java docs different for the sync API vs the async API. They should be the same. 

I think the above java doc makes more sense as compared to what we have on async API.

**Review by kushagraThapar — CHANGES_REQUESTED** [review]

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java [hunk 12]
```diff
@@ -1473,6 +1475,16 @@ public String getId() {
         return withContext(context -> deleteItemInternal(itemId, null, requestOptions, context));
     }
 
+    @Beta(value = Beta.SinceVersion.V4_19_0, warningText = Beta.PREVIEW_SUBJECT_TO_CHANGE_WARNING)
```
[the remark this answers is not in the document]
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:1488]: The above comment is old, please ignore it.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/BaseAuthorizationTokenProvider.java [hunk 13]
```diff
@@ -49,6 +49,7 @@ private static String getResourceSegment(ResourceType resourceType) {
         case Document:
             return Paths.DOCUMENTS_PATH_SEGMENT;
         case DocumentCollection:
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/BaseAuthorizationTokenProvider.java:51]: could you please check on the other auth modes as well? AAD, is this compatible with AAD as well?
@milismsft FYI

**Review by moderakh — APPROVED** [review]
as you are changing BaseAuthorizationTokenProvider.java 
please test if this is AAD compat

@milismsft FYI

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosContainer.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosContainer.java:645]: Added the same in async API as well

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:1490]: Added the same from the sync API

**Review by kushagraThapar — APPROVED** [review]

**Pull request merged**